### ISSUE TITLE
runtime/pprof: multithreaded CPU profiles incorrect on NetBSD

### ISSUE THREAD

**Issue by rsc** [golang/go#6047]
<pre>Multithreaded CPU profiles do not work on OS X, but worse they give little indication
that they are broken.
Single-threaded profiles kind of work, but only because we play games changing the
signal mask every time we start executing Go code, so we incur overhead on every thread
switch, and that in turn is suspected of causing deadlocks (see <a href="https://golang.org/issue/5519">issue #5519</a>). Also, many
times you can't quite be sure you hit the single-threaded case, so the profiles are not
truly believable. For years I have pulled out my Linux laptop when I want to do real
performance work with profiles.

The fundamental problem is that OS X delivers the profiling signals to the wrong thread
(golang.org/change/35b716c94225). The signal mask trick is an attempt to work around
this, but it's incomplete and misleading.

I intend to remove the workarounds from package runtime, so that CPU profiles on OS X
will typically be empty, because the wrong thread will receive the signals. That will
avoid giving out inaccurate information. It should also fix <a href="https://golang.org/issue/5519">issue #5519</a> and simplify the
scheduler.

The correct place to fix this bug is in the OS X kernel. I filed an Apple Bug Report in
2011 (again see golang.org/change/35b716c94225) but the problem remains.

For people who insist on accurate profiling on OS X and are a bit adventurous, it may be
possible to modify the kernel to deliver the profiling signal correctly. See
<a href="http://godoc.org/code.google.com/p/rsc/cmd/pprof_mac_fix">http://godoc.org/code.google.com/p/rsc/cmd/pprof_mac_fix</a> for details, and heed the
warnings in its documentation.</pre>

**Comment by adonovan** [comment]
Until this is fixed, can we make runtime.StartCPUProfiling print a message to stderr on Mac OS X to warn users that profiling is broken?

**Comment by rsc** [comment]
@adonovan, profiling works fine on OS X if you patch your kernel, as I think most Go programmers on Macs do, out of necessity. Certainly people running non-buggy kernels don't need to see forced output on standard error every time they profile a program.

If we had good way to tell whether the kernel patch has been applied, we could print a warning in that case. But I don't have a good way to do that. I have thought about changing the kernel version string but the only thing I am confident about changing is the date, and there's not much room there to signal that the fix is applied.

See rsc.io/pprof_mac_fix for the patch.

**Comment by pires** [comment]
@rsc thanks for the patch. Pray for me while I run it!

**Comment by rsc** [comment]
I have a report from an OS X 10.11 El Capitan beta user that pprof works out of the box on that system, without the need for a kernel patch. I have also inspected the machine code for the relevant kernel function, and they did make changes roughly along the lines of what the patch has always done. So I believe it was intentionally fixed. I am hopeful that the fix will last into the final public release of OS X 10.11 El Capitan. And then maybe years from now we can look back and laugh at how ridiculous it was that we had to apply a binary patch to our kernels to profile our programs.

Does anyone know: is NetBSD still broken?

**Comment by davecheney** [comment]
And there was much rejoicing and dancing in the streets.

**Comment by minux** [comment]
Judging by reading NetBSD kernel source code, I think the
system still traverse the thread list and picks the first thread
that accepts the signal to send the SIGPROF signal to.

The sigtest.c seems to confirm this finding with the majority
of the signal received by the last created thread.
$ ./sigtest
500 signals received.
0 on main
4 on looper 0
19 on looper 1
73 on looper 2
404 on looper 3

To verify my hypothesis, I added another sleeping thread to
sigtest.c created after all the other threads, and sure enough,
all the SIGPROFs are delivered to the last thread.
$ ./sigtest
0 on main
0 on looper 0
0 on looper 1
0 on looper 2
0 on looper 3
500 on sleep thread

This test is very reproducible.

In conclusion, multi-threaded profiling is still broken on NetBSD.

**Comment by jnjackins** [comment]
Confirming that profile does indeed work out-of-the-box on El Capitan.

**Comment by rsc** [comment]
Obsoleting in favor of #13841.

**Comment by alandonovan** [comment]
Yay!
